### PR TITLE
Issue #13345: Update scope attribute in membershipAttribute and member…

### LINKED ISSUE #13345: ldapRegistry-3.0 'scope' attributes do not specify valid values
The LDAP registry has several 'scope' attributes that do not define valid values. These attributes should either update the description to include valid values, or the values should be turned into enumerations that define these values.

### PR DESCRIPTION
**Pull request by jvanhill**
…Attribute to be an enum. Add FAT that tests the scopes. Update LDAPUtils to handle issue with older LDAP servers not negotiating the TLS connection from TLSv1.2/3 to TLSv1.

fixes #13345

Found (and opened) a few issues when writing a FAT for this change:

#13877
#13878
#14249

### CODE REVIEW

**Comment by jvanhill** [comment]
#build

**Review by ryanesch — APPROVED** [review]

**Review by helyarp — COMMENTED** [review]
Looks good. I'll add the ID reviewed label.

**Comment by jvanhill** [comment]
#build

**Comment by jvanhill** [comment]
The latest builds we are seeing the following two failures. The reason appears to be since I am modifying the attribute to be a string enumeration, the generated schema doesn't have a type that matches the last versions schema. To get around this, I will need to update the historical schema in `ant_schemaValidator` (in WS-CD-Open) to match what will be generated. Since we are so close to releasing 20.0.0.10, I need to wait until the `200010-schema.xsd` file is generated from the release so that I don't have to do this twice.

- com.ibm.ws.config.internal.schemaChecker.MetaTypeValidator
  - com.ibm.ws.security.wim.adapter.ldap.memberAttribute incorrectType
`junit.framework.AssertionFailedError: Type com.ibm.ws.security.wim.adapter.ldap.memberAttribute has attribute scope which should be of type xsd:string but is type .`

- com.ibm.ws.config.internal.schemaChecker.MetaTypeValidator
  - com.ibm.ws.security.wim.adapter.ldap.membershipAttribute incorrectType
`junit.framework.AssertionFailedError: Type com.ibm.ws.security.wim.adapter.ldap.membershipAttribute has attribute scope which should be of type xsd:string but is type .`

**Comment by jvanhill** [comment]
#build

**Comment by jvanhill** [comment]
Cancelled latest build request and re-requested using a custom `git.clone.checkout.branch` that contains changes to the 20.0.0.10 schema XSD. This should result in the errors in previous builds being eliminated.

https://wasrtc.hursley.ibm.com:9443/jazz/web/projects/WS-CD#action=com.ibm.team.build.viewDefinition&id=_JWdEUD7dEemKFI0xRbWJ_w

**Comment by jvanhill** [comment]
Needed to respin as I got the branch incorrect.

https://wasrtc.hursley.ibm.com:9443/jazz/web/projects/WS-CD#action=com.ibm.team.build.viewResult&id=_ZUmIQPyBEeqZosoAJ-GPIQ

**Comment by jvanhill** [comment]
Had an issue in the WS build, even though the previous build was good. Respinning with changes to fix WS build.

https://wasrtc.hursley.ibm.com:9443/jazz/web/projects/WS-CD#action=com.ibm.team.build.viewResult&id=_jIL04P0TEeqZosoAJ-GPIQ

**Comment by jvanhill** [comment]
Waiting for changes to schema made in WS-CD-Open to make it to master.

**Comment by jvanhill** [comment]
#build
#libby

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LDAPMemberAttributeScopeTest.java [hunk 1]
```diff
@@ -0,0 +1,577 @@
...
+
+        results = urServlet.getGroupsForUser(MEMBER_LEVEL2_USER_DN);
+        assertMembershipResults("With direct scope, the nested member should be a member of the root and nested group.", results, MEMBER_LEVEL1_GROUP_DN, MEMBER_LEVEL2_GROUP_DN);
+
+        results = urServlet.getGroupsForUser(MEMBER_DYNAMIC_USER_DN);
+        assertMembershipResults("With direct scope, the dynamic member should be a member of the root and dynamic group.", results, MEMBER_LEVEL1_GROUP_DN,
+                                MEMBER_DYNAMIC_GROUP_DN);
+
+        results = urServlet.getUsersForGroup(MEMBER_LEVEL1_GROUP_DN, 0).getList();
+        assertMembershipResults("With direct scope, the root group should have both the direct and nested members.", results, MEMBER_LEVEL1_USER_DN, MEMBER_LEVEL2_USER_DN);
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LDAPMemberAttributeScopeTest.java:366]: Why is dynamic group membership not resolved?

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LDAPMemberAttributeScopeTest.java [hunk 2]
```diff
@@ -0,0 +1,577 @@
...
+
+        List<String> results = urServlet.getGroupsForUser(MEMBERSHIP_LEVEL1_USER_DN);
+        assertMembershipResults("With direct scope, the direct member should only be a member of the root group.", results, MEMBERSHIP_LEVEL1_GROUP_DN);
+
+        results = urServlet.getGroupsForUser(MEMBERSHIP_LEVEL2_USER_DN);
+        assertMembershipResults("With direct scope, the nested member should be a member of the root and nested group.", results, MEMBERSHIP_LEVEL1_GROUP_DN,
+                                MEMBERSHIP_LEVEL2_GROUP_DN);
+
+        results = urServlet.getGroupsForUser(MEMBERSHIP_DYNAMIC_USER_DN);
+        assertMembershipResults("With direct scope, the dynamic member should be a member of the root and dynamic group.", results, MEMBERSHIP_DYNAMIC_GROUP_DN);
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LDAPMemberAttributeScopeTest.java:474]: Why is this not returning level 1 group?

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LDAPMemberAttributeScopeTest.java [hunk 3]
```diff
@@ -0,0 +1,577 @@
...
+        results = urServlet.getGroupsForUser(MEMBERSHIP_LEVEL2_USER_DN);
+        assertMembershipResults("With nested scope, the nested member should only be a member of the nested group.", results, MEMBERSHIP_LEVEL1_GROUP_DN,
+                                MEMBERSHIP_LEVEL2_GROUP_DN);
+
+        results = urServlet.getGroupsForUser(MEMBERSHIP_DYNAMIC_USER_DN);
+        assertMembershipResults("With nested scope, the dynamic member should be a member of the dynamic group.", results, MEMBERSHIP_DYNAMIC_GROUP_DN);
+
+        // TODO user_level2 is returned b/c recursiveSearch=true overrides search scope in LdapAdapter.getMembersByMembership()
+        results = urServlet.getUsersForGroup(MEMBERSHIP_LEVEL1_GROUP_DN, 0).getList();
+        assertMembershipResults("With nested scope, the root group should have the direct member.", results, MEMBERSHIP_LEVEL1_USER_DN, MEMBERSHIP_LEVEL2_USER_DN);
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LDAPMemberAttributeScopeTest.java:515]: Why is the dynamic user not returned?

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LDAPMemberAttributeScopeTest.java [hunk 4]
```diff
@@ -0,0 +1,577 @@
...
+        assertMembershipResults("With direct scope, the direct member should only be a member of the root group.", results, MEMBER_LEVEL1_GROUP_DN);
+
+        results = urServlet.getGroupsForUser(MEMBER_LEVEL2_USER_DN);
+        assertMembershipResults("With direct scope, the nested member should be a member of the root and nested group.", results, MEMBER_LEVEL1_GROUP_DN, MEMBER_LEVEL2_GROUP_DN);
+
+        results = urServlet.getGroupsForUser(MEMBER_DYNAMIC_USER_DN);
+        assertMembershipResults("With direct scope, the dynamic member should be a member of the root and dynamic group.", results, MEMBER_LEVEL1_GROUP_DN,
+                                MEMBER_DYNAMIC_GROUP_DN);
+
+        results = urServlet.getUsersForGroup(MEMBER_LEVEL1_GROUP_DN, 0).getList();
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LDAPMemberAttributeScopeTest.java:365]: The LdapAdapter.getMembersByMember() method does not consider nested dynamic groups.

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LDAPMemberAttributeScopeTest.java [hunk 5]
```diff
@@ -0,0 +1,577 @@
...
+
+        // TODO group_level1 is returned b/c recursiveSearch=true overrides search scope in LdapAdapter.getGroupsByMember()
+        results = urServlet.getGroupsForUser(MEMBER_LEVEL2_USER_DN);
+        assertMembershipResults("With all scope, the nested member should only be a member of the nested group.", results, MEMBER_LEVEL1_GROUP_DN, MEMBER_LEVEL2_GROUP_DN);
+
+        // TODO group_level1 is returned b/c recursiveSearch=true overrides search scope in LdapAdapter.getGroupsByMember()
+        results = urServlet.getGroupsForUser(MEMBER_DYNAMIC_USER_DN);
+        assertMembershipResults("With all scope, the dynamic member should be a member of the dynamic group.", results, MEMBER_LEVEL1_GROUP_DN, MEMBER_DYNAMIC_GROUP_DN);
+
+        // TODO user_level1 is returned b/c recursiveSearch=true overrides search scope in LdapAdapter.getMembersByMember()
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LDAPMemberAttributeScopeTest.java:440]: The dynamic user should be a member of the level 1 group, but the LdapAdapter.getMembersByMembership() method does not consider nested dynamic groups.

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LDAPMemberAttributeScopeTest.java [hunk 6]
```diff
@@ -0,0 +1,577 @@
...
+        assertMembershipResults("With direct scope, the direct member should only be a member of the root group.", results, MEMBERSHIP_LEVEL1_GROUP_DN);
+
+        results = urServlet.getGroupsForUser(MEMBERSHIP_LEVEL2_USER_DN);
+        assertMembershipResults("With direct scope, the nested member should be a member of the root and nested group.", results, MEMBERSHIP_LEVEL1_GROUP_DN,
+                                MEMBERSHIP_LEVEL2_GROUP_DN);
+
+        results = urServlet.getGroupsForUser(MEMBERSHIP_DYNAMIC_USER_DN);
+        assertMembershipResults("With direct scope, the dynamic member should be a member of the root and dynamic group.", results, MEMBERSHIP_DYNAMIC_GROUP_DN);
+
+        results = urServlet.getUsersForGroup(MEMBERSHIP_LEVEL1_GROUP_DN, 0).getList();
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LDAPMemberAttributeScopeTest.java:476]: The dynamic user should be a member of the level 1 group, but the LdapAdapter.getMembersByMembership() method does not consider nested dynamic groups.

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LDAPMemberAttributeScopeTest.java (hunk 1 above)
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LDAPMemberAttributeScopeTest.java:366]: The dynamic user should be a member of the level 1 group, but the LdapAdapter.getMembersByMember() method does not consider nested dynamic groups.

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LDAPMemberAttributeScopeTest.java [hunk 7]
```diff
@@ -0,0 +1,577 @@
...
+        assertMembershipResults("With direct scope, the root group should have both the direct and nested members.", results, MEMBERSHIP_LEVEL1_USER_DN, MEMBERSHIP_LEVEL2_USER_DN);
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LDAPMemberAttributeScopeTest.java:477]: The dynamic user should be a member of the level 1 group, but the LdapAdapter.getMembersByMembership() method does not consider nested dynamic groups.

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LDAPMemberAttributeScopeTest.java (hunk 1 above)
Comment (jvanhill) [reply dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LDAPMemberAttributeScopeTest.java:366]: Opened issue #14249.

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LDAPMemberAttributeScopeTest.java (hunk 7 above)
Comment (jvanhill) [reply dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LDAPMemberAttributeScopeTest.java:477]: Opened issue #14249.

**Comment by jvanhill** [comment]
Review found some comments needed to be updated and added to reference the issues that were opened to address those bugs. Since nothing but comments in the FAT changed from the latest build, I am not going to run a new build.

**Comment by jvanhill** [comment]
#libby

**Pull request merged**